platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -242,27 +255,75 @@ public void apply(final Project project) {
         project.getPluginManager().apply("org.gradle.jvm-test-suite"); // TODO: change to reference plugin class by name after project dependency cycles untangled; this will affect ApplyPluginBuildOperationIntegrationTest (will have to remove id)
 
-        // Create the 'java' component.
-        DefaultJvmSoftwareComponent component = objectFactory.newInstance(
-            DefaultJvmSoftwareComponent.class,
-            JvmConstants.JAVA_COMPONENT_NAME, SourceSet.MAIN_SOURCE_SET_NAME
-        );
-        project.getComponents().add(component);
+        JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+        DefaultJvmSoftwareComponent javaComponent = createJavaComponent(project, javaExtension, projectInternal);
 
         // Set the 'java' component as the project's default.
         Configuration defaultConfiguration = project.getConfigurations().getByName(Dependency.DEFAULT_CONFIGURATION);
-        defaultConfiguration.extendsFrom(component.getMainFeature().getRuntimeElementsConfiguration());
-        ((SoftwareComponentContainerInternal) project.getComponents()).getMainComponent().convention(component);
+        defaultConfiguration.extendsFrom(javaComponent.getMainFeature().getRuntimeElementsConfiguration());
+        ((SoftwareComponentContainerInternal) project.getComponents()).getMainComponent().convention(javaComponent);
 
-        JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
         BuildOutputCleanupRegistry buildOutputCleanupRegistry = projectInternal.getServices().get(BuildOutputCleanupRegistry.class);
         configureSourceSets(javaExtension, buildOutputCleanupRegistry);
 
         configureTestTaskOrdering(project.getTasks());
-        configureDiagnostics(project, component);
+        configureDiagnostics(project, javaComponent);
         configureBuild(project);
     }
 
+    private static DefaultJvmSoftwareComponent createJavaComponent(Project project, JavaPluginExtension javaExtension, ProjectInternal projectInternal) {
+        // Create the 'java' component - create sourceset first
+        SourceSet sourceSet = createSourceSet(SourceSet.MAIN_SOURCE_SET_NAME, javaExtension.getSourceSets());
+
+        // Supply the sourceSet to the feature
+        JvmFeatureInternal feature = new DefaultJvmFeature(
+            JvmConstants.JAVA_FEATURE_NAME, sourceSet, Collections.emptyList(),
+            (ProjectInternal) project, false, false);
+
+        // And supply main feature to the component
+        DefaultJvmSoftwareComponent component = project.getObjects().newInstance(
+            DefaultJvmSoftwareComponent.class,
+            JvmConstants.JAVA_COMPONENT_NAME,
+            project,
+            feature
+        );
+        project.getComponents().add(component);
+
+        // Then configure publishing on it
+        configurePublishing(project.getPlugins(), project.getExtensions(), sourceSet);
```

Comment:
💭 IMO this belongs outside of the `createJavaComponent` method. 

In practice, it sets defaults for _all_ publications. It just happens to source those defaults from the java component.

Reply:
Done.